Attempt at simplifying the file structure and CMake build

starseeker opened the pull request:
Collapses the various features into libmanifold, with configure options to enable/disable them.

Groups public headers in include/manifold.

Adjustments to the exported targets to try and avoid invoking the download steps in users of an installed Manifold library.

starseeker commented:
Follow-on from https://github.com/elalish/manifold/issues/945 and https://github.com/elalish/manifold/pull/953

starseeker commented:
@elalish   I'll pick back up with the EXPORT config logic on Monday... getting closer at least.

elalish commented on include/manifold/tri_dist.h:
What decides whether a header file is in `include` vs `src`? I was guessing only the public headers would go in `include` and the internal ones in `src`, but I think this is an example that's internal?

starseeker replied:
Correct - I got fooled by its use in the test files.  Just pushed a fix.

elalish commented on CMakeLists.txt:
```diff
@@ -15,31 +15,58 @@
-# fuzztest is a rather large dependency
-option(MANIFOLD_FUZZ "Enable fuzzing tests" OFF)
 option(MANIFOLD_DEBUG "Enable debug tracing/timing" OFF)
+option(MANIFOLD_DOWNLOADS "Allow Manifold build to download missing dependencies" ON) 
```
This is the new compile option, right? So this is the main new feature of this PR, beyond the major simplification of our stucture and CMake.

starseeker replied:
Yes - most of the time it will be a no-op, so it's not hugely significant, but it's handy for those projects that want to be sure their building of Manifold isn't downloading anything.

elalish commented on CMakeLists.txt:
```diff
@@ -140,75 +146,142 @@ if("${isSystemDir}" STREQUAL "-1")
+	message(STATUS "Environment: MXE")
 elseif (APPLE)
-  message(STATUS "Environment: macOS")
+	message(STATUS "Environment: macOS")
```
This is minor, but we generally do two-space indenting around here. Does clang-format handle CMake? For some reason ours has never been set up to run on these files, but it probably should be.

starseeker replied:
Hmm.  Good question.  A quick check with stackoverflow turns up https://github.com/cheshirekow/cmake_format and https://github.com/BlankSpruce/gersemi but I've not tried them to know how they work.  For BRL-CAD we usually just add a footer to help the editors:  https://github.com/BRL-CAD/brlcad/blob/f8d7be0f755ff0f332d5c25143fa06635343d57c/CMakeLists.txt#L2419

starseeker replied:
I committed an example in the pr reformatting the toplevel CMakeLists.txt the say way BRL-CAD does it - if that looks OK I can do the other CMake files similarly.

elalish commented on include/manifold/CMakeLists.txt:
```diff
@@ -0,0 +1,30 @@
+endif()
+
+if (MANIFOLD_EXPORT)
+	list(APPEND MANIFOLD_PUBLIC_HDRS meshIO.h)
```
Okay, so this is what we do instead of conditionally `addsubdirectory` - yeah that does seem cleaner. I kind of liked having the src files organized into a few directories, but this should be fine too.

starseeker replied:
If you want we can do subdirectories within src/ and just have the src/CMakeLists.txt file reference them there.

elalish replied:
Yeah sounds good. I think at the very least I'd like to separate `meshIO` since I really don't consider that to be a core part of Manifold. How about anything that's an optional component gets its own sub-directory?

starseeker replied:
OK, I put meshIO and cross_section files in subdirectories.

(FWIW - meshIO may not be a core part of Manifold, but since I can't properly prepare example failure cases without it it's fairly critical to me :-P)

starseeker replied:
Actually, that's a thought... what about maybe teaching core manifold to import and export just minimalist .glb files, since that's what's needed for debugging/testing, and have the assimp translation live as a test or sample code?

elalish replied:
In fact we have that, but it's in TS (gltf-io.ts) and it too depends on an external library. It's tricky because Manifold is designed to handle multiple materials, vertex properties, etc, so minimal I/O doesn't necessarily let us test everything either. Are you willing to help us maintain it?

starseeker replied:
Actually, I might be able to do what I need on my side, without it needing to go in manifold at all - that's probably the way to go.

elalish commented on include/manifold/conv.h:
```diff
@@ -17,18 +17,24 @@
 
 using namespace manifold;
 using ManifoldVec = std::vector<Manifold>;
+#ifdef MANIFOLD_CROSS_SECTION
```
I think we can skip this if we continue to say the bindings always require Clipper2. Or was someone wanting C bindings without cross-section?

starseeker replied:
I did that because I'm always building the C bindings as part of the main libmanifold, so they need to conditionalize based on whether the user designates cross_section as on or off.  It might be a little cleaner to break the C binding logic specific to cross_section out into a separate file that can be turned on or off in the build logic, if you prefer.

elalish replied:
That sounds like a good idea, but I'll defer to your judgement.

starseeker replied:
I played with it a little, and based on that for the moment I'd prefer to leave things as is, since it's compiling and working.  We can always refine it later.

elalish reviewed (commented):
Thanks for putting this all together - and great to see it passing our CI. @kintel @fire @wtholliday please put your two cents in here too - would be nice to not have to redo this again, and I don't have a strong opinion on structure besides making it easy to consume downstream.

wtholliday commented:
> please put your two cents in here too

I'm not qualified to evaluate the cmake stuff, but the directory structure makes me happy :)

pca006132 commented on sh/format.sh:
I think we should name the directory `scripts`?

starseeker replied:
Sure - sh was just the BRL-CAD convention, but scripts is more generic.

pca006132 commented on manifoldConfig.cmake.in:
```diff
@@ -1,5 +1,30 @@
+# for bundled dependencies.
+get_filename_component(_FIND_ROOT "${CMAKE_CURRENT_LIST_FILE}" PATH)
+get_filename_component(_FIND_ROOT "${_FIND_ROOT}" PATH)
+get_filename_component(_FIND_ROOT "${_FIND_ROOT}" PATH)
```
are these duplicated lines intended or a mistake?

starseeker replied:
Intended - it's constructing a parent path by stripping off the children.  The CMAKE_CURRENT_LIST_FILE is a couple levels down from the path the script wants, so it does multiple calls to walk "up" the path.

pca006132 replied:
I see. Maybe add this as a comment?

pca006132 reviewed (commented):
btw maybe we should take the chance to change `MANIFOLD_PAR` to ON/OFF as well, we no longer have multiple backends anyway

pca006132 commented on src/CMakeLists.txt:
```diff
@@ -12,10 +12,142 @@
+		endif()
+	else()
+		if(oneDPL_FOUND)
+			list(APPEND MANIFOLD_LIBS oneDPL)
```
btw I think we don't really need oneDPL, we just forgot to remove it.

pca006132 commented on CMakeLists.txt:
```diff
@@ -140,75 +146,142 @@ if("${isSystemDir}" STREQUAL "-1")
+	find_package(Threads REQUIRED)
+	find_package(TBB QUIET)
+	if(APPLE)
+		find_package(oneDPL QUIET)
```
oneDPL not really needed

starseeker commented:
> btw maybe we should take the chance to change `MANIFOLD_PAR` to ON/OFF as well, we no longer have multiple backends anyway

Do you want to make it binary just in CMake, or also in the C++ source code?  (i.e. #if MANIFOLD_PAR == 'T' -> #ifdef MANIFOLD_PAR)

pca006132 commented:
> > btw maybe we should take the chance to change `MANIFOLD_PAR` to ON/OFF as well, we no longer have multiple backends anyway
> 
> Do you want to make it binary just in CMake, or also in the C++ source code? (i.e. #if MANIFOLD_PAR == 'T' -> #ifdef MANIFOLD_PAR)

Both.

starseeker commented:
By the way, should that .vscode directory be committed into the repo?

pca006132 commented:
> By the way, should that .vscode directory be committed into the repo?

IMO no, I think people often use their own vscode config for the project anyway. @elalish what do you think?

wtholliday commented:
> Is it bothering anyone?

I thought it was nice, because I'm not a regular vscode user (Xcode usually) and vscode automatically did the code formatting, which I assume is part of your config. Only issue I had was that cmake fails within vscode on my mac, which perhaps is a config issue?

```
[cmake] CMake Error: Error: generator : Ninja
[cmake] Does not match the generator used previously: Unix Makefiles
[cmake] Either remove the CMakeCache.txt file and CMakeFiles directory or choose a different binary directory.
```

I blow away the whole build dir and get the same error. Works outside of vscode.

elalish commented:
@wtholliday Interesting, I'm not aware of Ninja being part of VSCode by default, but maybe something has changed?

starseeker commented on CMakeLists.txt:
```diff
@@ -311,39 +308,46 @@ message(STATUS " ")
+# mode: cmake
+# indent-tabs-mode: t
+# End:
+# ex: shiftwidth=2 tabstop=8
```
I can just format without adding the settings in the file if you'd rather not include something like that - I just wanted to offer that as an option to help folks keep the CMake formatting in line with Manifold's convention, since (at the moment) there isn't an automated tool enforcing it.

elalish replied:
Sure, that's fine by me - thanks for the explanation. I have almost never used vim or emacs, and sincerely hope I never need to.

starseeker commented:
About vec_view.h and iters.h - the former is referenced by cross_section.h, and the latter by parallel.h - that's why they got categorized as public headers, since cross_section.h and parallel.h are public?  I can check if they're actually needed in those headers...

starseeker commented:
Is cross_section.h intended to be a public header?  If so, it looks like vec_view.h needs to be too...

elalish commented:
> Is cross_section.h intended to be a public header? If so, it looks like vec_view.h needs to be too...

Ah, I see. Yeah, parallel.h became public recently and I didn't realize it pulled in iter.h. And vec_view apparently came in to both cross_section and manifold.h with WarpBatch. Okay, I guess that's fine as is, thanks.

elalish commented:
Do we think this is ready to merge? If so I can update our CI so it's expecting these test runs.

starseeker commented:
> Ah, I see. Yeah, parallel.h became public recently and I didn't realize it pulled in iter.h. And vec_view apparently came in to both cross_section and manifold.h with WarpBatch. Okay, I guess that's fine as is, thanks.

Oh, whoops - I make parallel.h non-public to hide iter.h.  Should I put it back?

starseeker commented:
> > Is cross_section.h intended to be a public header? If so, it looks like vec_view.h needs to be too...
> 
> Ah, I see. Yeah, parallel.h became public recently and I didn't realize it pulled in iter.h. And vec_view apparently came in to both cross_section and manifold.h with WarpBatch. Okay, I guess that's fine as is, thanks.

OK, I put iter.h and parallel.h back.  As far as I know this can be merged, as long as nobody else has spotted any issues.  It'll probably be easier to do any additional refinements in other pull requests in any case.

starseeker commented:
Rebased, corrected indenting on other CMakeLists.txt files.  Assuming the adjustment to CMake spacing didn't inadvertently break something, that should be all I've got for this PR.

starseeker commented:
> Thanks - really glad to have your help cleaning up our CMake and structure.

My pleasure - thank you for being willing to do so!  It makes life much easier for me updating to newer Manifold versions (I was doing some custom modding on my end prior to this) so it's a textbook win-win.
